nameko/testing/services.py

```diff
@@ -48,93 +48,157 @@ def entrypoint_hook(container, method_name, context_data=None):
+            entrypoint raised an exception, otherwise `None`.
 
-    waiter = EntrypointWaiter(entrypoint)
-    if not get_extension(container, Entrypoint, method_name=entrypoint):
+
+    The `timeout` argument specifies the maximum number of seconds the
+    `entrypoint_waiter` should wait before exiting. It can be disabled by
+    passing `None`.
+
+    **Usage**
+
+    ::
+        class Service(object):
+            name = "service"
+
+            @event_handler('srcservice', 'eventtype')
+            def handle_event(self, msg):
+                return msg
+
+        container = ServiceContainer(Service, config)
+        container.start()
+
+        # basic
+        with entrypoint_waiter(container, 'handle_event'):
+            ...  # action that dispatches event
+
+        # giving access to the result
+        with entrypoint_waiter(container, 'handle_event') as result:
+            ...  # action that dispatches event
+        res = result.get()
+
+        # with custom timeout
+        with entrypoint_waiter(container, 'handle_event', timeout=5):
+            ...  # action that dispatches event
+
+        # with callback that waits until entrypoint stops raising
+        def callback(worker_ctx, result, exc_info):
+            if exc_info is None:
+                return True
+
+        with entrypoint_waiter(container, 'handle_event', callback=callback):
+            ...  # action that dispatches event
+
+    """
+    if not get_extension(container, Entrypoint, method_name=method_name):
         raise RuntimeError("{} has no entrypoint `{}`".format(
-            container.service_name, entrypoint))
-    if get_extension(container, EntrypointWaiter, entrypoint=entrypoint):
-        raise RuntimeError("Waiter already registered for {}".format(
-            entrypoint))
-
-    # can't mess with dependencies while container is running
-    wait_for_worker_idle(container)
-    container.dependencies.add(waiter)
-
-    try:
-        yield
-        exc = waiter.Timeout(
-            "Entrypoint {}.{} failed to complete within {} seconds".format(
-                container.service_name, entrypoint, timeout)
-        )
-        with eventlet.Timeout(timeout, exception=exc):
-            waiter.wait()
-    finally:
-        wait_for_worker_idle(container)
-        container.dependencies.remove(waiter)
+            container.service_name, method_name))
+
+    waiter_callback = callback
+    waiter_result = WaitResult()
+
+    def on_worker_result(worker_ctx, result, exc_info):
+        complete = False
+        if worker_ctx.entrypoint.method_name == method_name:
+            if not callable(waiter_callback):
+                complete = True
+            else:
+                complete = waiter_callback(worker_ctx, result, exc_info)
+
+        if complete:
+            waiter_result.send(result, exc_info)
+        return complete
+
+    def on_worker_teardown(worker_ctx):
+        if worker_ctx.entrypoint.method_name == method_name:
```

Comment:
actually. could this become a race with another call with the same method name for which the callback returned false? could this e.g. check if it was called with the same object (`is`) as on_worker_result?